docs/static/field-reference.asciidoc

```diff
@@ -0,0 +1,129 @@
+      # +----string-argument---+    |  +--field-ref----+    +--field-ref----+ |
+      #                             +-------- sprintf format string ----------+
+      }
+    }
+  }
+}
+
+[[formal-grammar]]
+=== Formal Grammar
+
+Below is the formal grammar of the Field Reference, with notes and examples.
+
+[[formal-grammar-field-reference-literal]]
+==== Field Reference Literal
+
+A _Field Reference Literal_ is a sequence of one or more _Path Fragments_ that can be used directly in Logstash pipeline <<configuration#conditionals,conditionals>> without any additional quoting (e.g. `[request]`, `[response][status]`).
+
+[source,antlr]
+fieldReferenceLiteral
+  : ( pathFragment )+
+  ;
+
+[[formal-grammar-field-reference]]
+==== Field Reference (Event APIs)
+
+The Event API's methods for manipulating the fields of an event or using the sprintf syntax are more flexible than the pipeline grammar in what they accept as a Field Reference.
+Top-level fields can be referenced directly by their _Field Name_ without the square brackets, and there is some support for _Composite Field References_, simplifying use of programmatically-generated Field References.
+
+A _Field Reference_ for use with the Event API is therefore one of:
+
+ - a single _Field Reference Literal_; OR
+ - a single _Field Name_ (referencing a top-level field); OR
+ - a single _Composite Field Reference_.
+
+[source,antlr]
+eventApiFieldReference
+  : fieldReferenceLiteral
+  | fieldName
+  | compositeFieldReference
+  ;
+
+[[formal-grammar-path-fragment]]
+==== Path Fragment
+
+A _Path Fragment_ is a _Field Name_ wrapped in square brackets (e.g., `[request]`).
+
+[source,antlr]
+pathFragment
+  : '[' fieldName ']'
+  ;
+
+[[formal-grammar-field-name]]
+==== Field Name
+
+A _Field Name_ is a sequence of characters that are _not_ square brackets (`[` or `]`).
+
+[source,antlr]
+fieldName
+  : ( ~( '[' | ']' ) )+
+  ;
+
+[[formal-grammar-event-api-composite-field-reference]]
+==== Composite Field Reference
+
+In some cases, it may be necessary to programmatically _compose_ a Field Reference from one or more Field References,
+such as when manipulating fields in a plugin or while using the Ruby Filter plugin and the Event API.
+
+[source,ruby,indent=4]
+    fieldReference = "[path][to][deep nested field]"
+    compositeFieldReference = "[@metadata][#{fieldReference}][size]"
+    # => "[@metadata][[path][to][deep nested field]][size]"
+
+// NOTE: escapes are necessary to get `[[` sequences in .asciidoc tables to render properly in HTML.
+//       in each such case below, the unescaped input is noted in an asciidoc comment.
+.Canonical Representations of Composite Field References
+|===
+|Acceptable _Composite Field Reference_ |Canonical _Field Reference_ Representation
+
+|`[[deep][nesting]][field]`
+|`[deep][nesting][field]`
+
+// `[foo][[bar]][bingo]`
+|`[foo]\[[bar]][bingo]`
+|`[foo][bar][bingo]`
+
+// `[[ok]]`
+|`\[[ok]]`
+|`[ok]`
+|===
+
```

Comment:
It appears that asciidoc doesn't like something about the table.  Here's the error: 
`asciidoc: WARNING: field-reference.asciidoc: line 109: nested inline passthrough`

I'll keep experimenting to see if I can make it work.  Ideas are welcome.

Reply:
Argh. Okay. I'll just add a second table in comments that is wholly separate.

Reply:
I think using the asciidoc "plus for passthrough" syntax to escape the entire strings instead of randomly injecting backslashes reads clearly enough that we don't need the individual comments.